File: dart/test/tree_regexp_test.dart
import 'package:cucumber_expressions/src/tree_regexp.dart';
import 'package:test/test.dart';

void main() {
  group('TreeRegexp', () {
    test('exposes group source', () {
      final tr = TreeRegexp(RegExp('(a(?:b)?)(c)'));
      expect(
        tr.groupBuilder.children.map((gb) => gb.source).toList(),
        equals(['a(?:b)?', 'c']),
      );
    });

    test('builds tree', () {
      final tr = TreeRegexp(RegExp('(a(?:b)?)(c)'));
      final group = tr.match('ac')!;
      expect(group.value, equals('ac'));
      expect(group.children![0].value, equals('a'));
      expect(group.children![0].children, isNull);
      expect(group.children![1].value, equals('c'));
    });

    test('ignores ?: as a non-capturing group', () {
      final tr = TreeRegexp(RegExp('a(?:b)(c)'));
      final group = tr.match('abc')!;
      expect(group.value, equals('abc'));
      expect(group.children!.length, equals(1));
    });

    test('ignores ?! as a non-capturing group', () {
      final tr = TreeRegexp(RegExp('a(?!b)(.+)'));
      final group = tr.match('aBc')!;
      expect(group.value, equals('aBc'));
      expect(group.children!.length, equals(1));
    });

    test('ignores ?= as a non-capturing group', () {
      final tr = TreeRegexp(RegExp('a(?=[b])(.+)'));
      final group = tr.match('abc')!;
      expect(group.value, equals('abc'));
      expect(group.children!.length, equals(1));
      expect(group.children![0].value, equals('bc'));
    });

    test('ignores ?<= as a non-capturing group', () {
      final tr = TreeRegexp(RegExp(r'a(.+)(?<=c)$'));
      final group = tr.match('abc')!;
      expect(group.value, equals('abc'));
      expect(group.children!.length, equals(1));
      expect(group.children![0].value, equals('bc'));
    });

    test('ignores ?<! as a non-capturing group', () {
      final tr = TreeRegexp(RegExp(r'a(.+?)(?<!b)$'));
      final group = tr.match('abc')!;
      expect(group.value, equals('abc'));
      expect(group.children!.length, equals(1));
      expect(group.children![0].value, equals('bc'));
    });

    test('matches named capturing group', () {
      final tr = TreeRegexp(RegExp('a(?<name>b)c'));
      final group = tr.match('abc')!;
      expect(group.value, equals('abc'));
      expect(group.children!.length, equals(1));
      expect(group.children![0].value, equals('b'));
    });

    test('matches optional group', () {
      final tr = TreeRegexp(RegExp('^Something( with an optional argument)?'));
      final group = tr.match('Something')!;
      expect(group.children![0].value, isNull);
    });

    test('matches nested groups', () {
      final tr = TreeRegexp(RegExp(
          r'^A (\d+) thick line from ((\d+),\s*(\d+),\s*(\d+)) to ((\d+),\s*(\d+),\s*(\d+))'));
      final group = tr.match('A 5 thick line from 10,20,30 to 40,50,60')!;
      expect(group.children![0].value, equals('5'));
      expect(group.children![1].value, equals('10,20,30'));
      expect(group.children![1].children![0].value, equals('10'));
      expect(group.children![1].children![1].value, equals('20'));
      expect(group.children![1].children![2].value, equals('30'));
      expect(group.children![2].value, equals('40,50,60'));
      expect(group.children![2].children![0].value, equals('40'));
      expect(group.children![2].children![1].value, equals('50'));
      expect(group.children![2].children![2].value, equals('60'));
    });

    test('detects multiple non capturing groups', () {
      final tr = TreeRegexp(RegExp(r'(?:a)(:b)(\?c)(d)'));
      final group = tr.match('a:b?cd')!;
      expect(group.children!.length, equals(3));
    });

    test('works with escaped backslash', () {
      final tr = TreeRegexp(RegExp(r'foo\\(bar|baz)'));
      final group = tr.match('foo\\bar')!;
      expect(group.children!.length, equals(1));
    });

    test('works with escaped slash', () {
      final tr = TreeRegexp(RegExp(r"^I go to '\/(.+)'$"));
      final group = tr.match("I go to '/hello'")!;
      expect(group.children!.length, equals(1));
    });

    test('works with digit and word', () {
      final tr = TreeRegexp(RegExp(r'^(\d) (\w+)$'));
      final group = tr.match('2 you')!;
      expect(group.children!.length, equals(2));
    });

    test('captures non capturing groups with capturing groups inside', () {
      final tr = TreeRegexp.fromString('the stdout(?: from "(.*?)")?');
      final group = tr.match('the stdout')!;
      expect(group.value, equals('the stdout'));
      expect(group.children![0].value, isNull);
      expect(group.children!.length, equals(1));
    });

    test('works with case insensitive flag', () {
      final tr = TreeRegexp(RegExp('HELLO', caseSensitive: false));
      final group = tr.match('hello')!;
      expect(group.value, equals('hello'));
    });

    test('empty capturing group', () {
      final tr = TreeRegexp(RegExp('()'));
      final group = tr.match('')!;
      expect(group.value, equals(''));
      expect(group.children!.length, equals(1));
    });

    test('empty look ahead', () {
      final tr = TreeRegexp(RegExp('(?<=)'));
      final group = tr.match('')!;
      expect(group.value, equals(''));
      expect(group.children, isNull);
    });

    test('does not consider parenthesis in character class as group', () {
      final tr = TreeRegexp(RegExp(r'^drawings: ([A-Z, ()]+)$'));
      final group = tr.match('drawings: ONE(TWO)')!;
      expect(group.value, equals('drawings: ONE(TWO)'));
      expect(group.children!.length, equals(1));
      expect(group.children![0].value, equals('ONE(TWO)'));
    });
  });
}

File: elixir/test/varar/cucumber_expressions/tree_regexp_test.exs
defmodule Varar.CucumberExpressions.TreeRegexpTest do
  use ExUnit.Case, async: true

  alias Varar.CucumberExpressions.{Error, TreeRegexp}

  test "exposes the group source" do
    tr = TreeRegexp.new!(~r/(a(?:b)?)(c)/)
    assert Enum.map(tr.group_builder.children, & &1.source) == ["a(?:b)?", "c"]
  end

  test "builds a tree" do
    tr = TreeRegexp.new!(~r/(a(?:b)?)(c)/)
    group = TreeRegexp.match(tr, "ac")
    assert group.value == "ac"
    assert Enum.at(group.children, 0).value == "a"
    assert Enum.at(group.children, 0).children == nil
    assert Enum.at(group.children, 1).value == "c"
  end

  test "ignores `?:` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(?:b)(c)/)
    group = TreeRegexp.match(tr, "abc")
    assert group.value == "abc"
    assert length(group.children) == 1
    assert Enum.at(group.children, 0).value == "c"
  end

  test "ignores `?!` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(?!b)(.+)/)
    group = TreeRegexp.match(tr, "aBc")
    assert group.value == "aBc"
    assert length(group.children) == 1
  end

  test "ignores `?=` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(?=b)(.+)$/)
    group = TreeRegexp.match(tr, "abc")
    assert group.value == "abc"
    assert Enum.at(group.children, 0).value == "bc"
    assert length(group.children) == 1
  end

  test "ignores `?<=` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(.+)(?<=c)$/)
    group = TreeRegexp.match(tr, "abc")
    assert group.value == "abc"
    assert Enum.at(group.children, 0).value == "bc"
    assert length(group.children) == 1
  end

  test "ignores `?<!` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(.+)(?<!b)$/)
    group = TreeRegexp.match(tr, "abc")
    assert group.value == "abc"
    assert Enum.at(group.children, 0).value == "bc"
    assert length(group.children) == 1
  end

  test "ignores `?>` as a non-capturing group" do
    tr = TreeRegexp.new!(~r/a(?>b)c/)
    group = TreeRegexp.match(tr, "abc")
    assert group.value == "abc"
    assert group.children == nil
  end

  test "raises for named capture groups" do
    assert_raise Error, ~r/Named capture groups are not supported/, fn ->
      TreeRegexp.new!("^I am a person( named \"(?<first_name>.+) (?<last_name>.+)\")?$")
    end
  end

  test "matches an optional group" do
    tr = TreeRegexp.new!(~r/^Something( with an optional argument)?/)
    group = TreeRegexp.match(tr, "Something")
    assert Enum.at(group.children, 0).value == nil
  end

  test "matches nested groups" do
    tr =
      TreeRegexp.new!(
        ~r/^A (\d+) thick line from ((\d+),\s*(\d+),\s*(\d+)) to ((\d+),\s*(\d+),\s*(\d+))/
      )

    group = TreeRegexp.match(tr, "A 5 thick line from 10,20,30 to 40,50,60")

    assert Enum.at(group.children, 0).value == "5"
    assert Enum.at(group.children, 1).value == "10,20,30"
    assert Enum.map(Enum.at(group.children, 1).children, & &1.value) == ["10", "20", "30"]
    assert Enum.at(group.children, 2).value == "40,50,60"
    assert Enum.map(Enum.at(group.children, 2).children, & &1.value) == ["40", "50", "60"]
  end

  test "detects multiple non-capturing groups" do
    tr = TreeRegexp.new!(~r/(?:a)(:b)(\?c)(d)/)
    group = TreeRegexp.match(tr, "a:b?cd")
    assert length(group.children) == 3
  end

  test "works with escaped backslashes" do
    tr = TreeRegexp.new!(~r/foo\\(bar|baz)/)
    group = TreeRegexp.match(tr, "foo\\bar")
    assert length(group.children) == 1
  end

  test "works with escaped slashes" do
    tr = TreeRegexp.new!(~r/^I go to '\/(.+)'$/)
    group = TreeRegexp.match(tr, "I go to '/hello'")
    assert length(group.children) == 1
  end

  test "works with digit and word regexp metacharacters" do
    tr = TreeRegexp.new!(~r/^(\d) (\w+)$/)
    group = TreeRegexp.match(tr, "2 you")
    assert length(group.children) == 2
  end

  test "captures non-capturing groups with capturing groups inside" do
    tr = TreeRegexp.new!(~r/the stdout(?: from "(.*?)")?/)
    group = TreeRegexp.match(tr, "the stdout")
    assert group.value == "the stdout"
    assert Enum.at(group.children, 0).value == nil
    assert length(group.children) == 1
  end

  test "works with flags" do
    tr = TreeRegexp.new!(~r/HELLO/i)
    group = TreeRegexp.match(tr, "hello")
    assert group.value == "hello"
  end

  test "does not consider parentheses in regexp character classes as a group" do
    tr = TreeRegexp.new!(~r/^drawings: ([A-Z_, ()]+)$/)
    group = TreeRegexp.match(tr, "drawings: ONE, TWO(ABC)")
    assert group.value == "drawings: ONE, TWO(ABC)"
    assert Enum.at(group.children, 0).value == "ONE, TWO(ABC)"
    assert length(group.children) == 1
  end

  test "works with inline flags" do
    tr = TreeRegexp.new!("(?i)HELLO")
    group = TreeRegexp.match(tr, "hello")
    assert group.value == "hello"
    assert group.children == nil
  end

  test "works with non-capturing inline flags" do
    tr = TreeRegexp.new!("(?i:HELLO)")
    group = TreeRegexp.match(tr, "hello")
    assert group.value == "hello"
    assert group.children == nil
  end

  test "works with empty capturing groups" do
    tr = TreeRegexp.new!(~r/()/)
    group = TreeRegexp.match(tr, "")
    assert group.value == ""
    assert Enum.at(group.children, 0).value == ""
    assert length(group.children) == 1
  end

  test "works with empty non-capturing groups" do
    tr = TreeRegexp.new!(~r/(?:)/)
    group = TreeRegexp.match(tr, "")
    assert group.value == ""
    assert group.children == nil
  end

  test "works with empty non-look ahead groups" do
    tr = TreeRegexp.new!(~r/(?<=)/)
    group = TreeRegexp.match(tr, "")
    assert group.value == ""
    assert group.children == nil
  end
end
